front: setting power restrictions don't remove pathSteps

clarani opened the pull request:
closes #9361 

When manipulating power restrictions, the pathSteps without any constraints (stop, arrival time or margin) were cleaned. This was the expected behaviour before, because we could not have pathSteps without constraints.

As the expected behaviour changed (we now authorize pathSteps without constraints), the pathSteps created from the power restrictions selector have a new flag and may be cleaned if they are not linked anymore to a power restriction and have no constraints. The other pathSteps will not be cleaned.

:warning: There is currently another bug which resets automatically the power restrictions when editing a train (#9902). It should be fixed in another PR (#9924).

To test:
- create a train between PE and SG (select a rolling stock with power restrictions)
- add some via on the path (don't add any stop, arrival times or margins)
- add some power restrictions: cut a range in 2 and select a value
- check in the itinerary tab that a new via has appeared (or 2) for the power restriction range, but the other vias should still be there
- remove the power restriction range
- check in the itinerary tab that the vias linked to the power restriction range have disappeared, but not the other ones.

emersion commented on front/src/reducers/osrdconf/operationalStudiesConf/utils.ts:
```diff
@@ -83,7 +83,7 @@ export const addPathStep = (pathSteps: PathStep[], newPathStep: PathStep): PathS
       (step) => step.positionOnPath && step.positionOnPath > newPathStep.positionOnPath!
     );
-    return addElementAtIndex(pathSteps, index, newPathStep);
+    return addElementAtIndex(pathSteps, index, { ...newPathStep, isFromPowerRestriction: true });
```
This flag will be set when the path step is created, but when reloading the page this won't work anymore right?

clarani replied:
Yes, it won't work anymore, but it's the best way I found to avoid removing by mistake pathSteps which should not be removed :/

emersion replied:
This will only result in false negatives, ie. after page reload path steps related to power restrictions won't be cleared. Since there is no data loss, that's fine.

RomainValls reviewed (commented):
Found a little bug : 
Try cutting the range twice to add 2 power restrictions (ex: 0-50k ; 50k - 100k)
Cut the second restriction in two (like at 75k)
Check the pathsteps, everything is ok for now. 
Delete the newly created power restriction (the one from 50k to 75k) 
Delete the second power restriction (75k - 100k) 
We can still see the via at 75k

clarani commented:
You're right @RomainValls, thanks !
I forgot to add the flag when adding a new pathStep by "cutting a range", it's fixed now ;)

emersion commented on front/src/applications/operationalStudies/hooks/useManageTrainScheduleContext.tsx:
```diff
@@ -29,11 +34,27 @@ export const ManageTrainScheduleContextProvider = ({
     [pathProperties]
   );
 
+  useEffect(() => {
+    if (pathProperties) {
+      const allWaypoints = upsertPathStepsInOPs(
+        pathProperties.suggestedOperationalPoints,
+        compact(pathSteps)
+      );
+      setPathProperties({
+        ...pathProperties,
+        allWaypoints,
+      });
+    }
+  }, [pathSteps]);
```
Why is this necessary now? What does this fix?

clarani replied:
AllWaypoints is used to compute the inputs table.
Without this useEffect, if I add some vias from the power restrictions selector, then they are not added in allWaypoints and not displayed in the inputs table.

Note: I think it would be better to have a useMemo for allWaypoints, rather than to store them artificially in pathProperties

emersion replied:
Agreed! 

(Side note, this `useEffect()` depends on `pathProperties.suggestedOperationalPoints`, but that's missing from the dependencies.)

emersion replied:
> Note: I think it would be better to have a useMemo for allWaypoints, rather than to store them artificially in pathProperties

Done in https://github.com/OpenRailAssociation/osrd/pull/10169

emersion commented on front/src/reducers/osrdconf/operationalStudiesConf/utils.ts:
```diff
@@ -83,7 +83,7 @@ export const addPathStep = (pathSteps: PathStep[], newPathStep: PathStep): PathS
       (step) => step.positionOnPath && step.positionOnPath > newPathStep.positionOnPath!
     );
-    return addElementAtIndex(pathSteps, index, newPathStep);
+    return addElementAtIndex(pathSteps, index, { ...newPathStep, isFromPowerRestriction: true });
```
I'm a bit antsy to set this flag in a generic `addPathStep()` function… Could we maybe set it from the caller side?

clarani replied:
Done

emersion commented on front/src/reducers/osrdconf/operationalStudiesConf/utils.ts:
```diff
@@ -95,7 +95,7 @@ export const cleanPathSteps = (
 ): PathStep[] =>
   pathSteps.reduce((acc, pathStep, index) => {
-    if (index === 0 || index === pathSteps.length - 1) {
+    if (index === 0 || index === pathSteps.length - 1 || !pathStep.isFromPowerRestriction) {
```
Can we keep a TODO documenting the limitations of this approach?

clarani replied:
It's done :+1:

clarani committed:
front: setting power restrictions do not remove PathSteps

When manipulating power restrictions, the pathSteps without any constraints (stop, arrival time or margin) were cleaned.
This was the expected behaviour before, because we could not have pathSteps without constraints.

As the expected behaviour has changed (we now authorize pathSteps without constraints), the pathSteps created from the power restrictions selector
have a new flag and may be cleaned if they are not linked anymore to a power restrictions and have no constraints.
The other pathSteps will not be cleaned.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

emersion commented on front/src/reducers/osrdconf/types.ts:
```diff
@@ -89,6 +89,14 @@ export type PathStep = PathItemLocation & {
+   *
+   * If true, the pathStep might be cleaned if its power restriction is removed (except if it has time, stop or margin constraints)
+   *
+   * This flag will only work if the user has not saved its change. Once the change is saved, the flag will be removed and the pathStep
```
Nit: the "its" pronoun is not used for words referring to people, such as "the user". "[their](https://en.wikipedia.org/wiki/Singular_they)" can be used instead.
